### PR TITLE
[new release] ocaml-lsp-server, lsp and jsonrpc (1.8.0)

### PR DESCRIPTION
**Pull request by rgrinberg**
LSP Server for OCaml

- Project page: <a href="https://github.com/ocaml/ocaml-lsp">https://github.com/ocaml/ocaml-lsp</a>

## Features

- Add a new code action `Add missing rec keyword`, which is available when
  adding a `rec` keyword can fix `Unbound value ...` error, e.g.,

  ```ocaml
  let fact n = if n = 0 then 1 else n * fact (n - 1)
                                     (* ^^^^ Unbound value fact *)
  ```

  Adding `rec` to the definition of `fact` will fix the problem. The new code
  action offers adding `rec`.

- Use ocamlformat to properly format type snippets. This feature requires the
  `ocamlformat-rpc` opam package to be installed. (ocaml/ocaml-lsp#386)

- Add completion support for polymorphic variants, when it is possible to pin
  down the precise type. Examples (`<|>` stands for the cursor) when completion
  will work (ocaml/ocaml-lsp#473)

  Function application:

  ```
  let foo (a: [`Alpha | `Beta]) = ()

  foo `A<|>
  ```

  Type explicitly shown:

  ```
  let a : [`Alpha | `Beta] = `B<|>
  ```

  Note: this is actually a bug fix, since we were ignoring the backtick when
  constructing the prefix for completion.

- Parse merlin errors (best effort) into a more structured form. This allows
  reporting all locations as "related information" (ocaml/ocaml-lsp#475)

- Add support for Merlin `Construct` command as completion suggestions, i.e.,
  show complex expressions that could complete the typed hole. (ocaml/ocaml-lsp#472)

- Add a code action `Construct an expression` that is shown when the cursor is
  at the end of the typed hole, i.e., `_|`, where `|` is the cursor. The code
  action simply triggers the client (currently only VS Code is supported) to
  show completion suggestions. (ocaml/ocaml-lsp#472)

- Change the formatting-on-save error notification to a warning notification
  (ocaml/ocaml-lsp#472)

- Code action to qualify ("put module name in identifiers") and unqualify
  ("remove module name from identifiers") module names in identifiers (ocaml/ocaml-lsp#399)

  Starting from:

  ```ocaml
  open Unix

  let times = Unix.times ()
  let f x = x.Unix.tms_stime, x.Unix.tms_utime
  ```

  Calling "remove module name from identifiers" with the cursor on the open
  statement will produce:

  ```ocaml
  open Unix

  let times = times ()
  let f x = x.tms_stime, x.tms_utime
  ```

  Calling "put module name in identifiers" will restore:

  ```ocaml
  open Unix

  let times = Unix.times ()
  let f x = x.Unix.tms_stime, x.Unix.tms_utime
  ```

## Fixes

- Do not show "random" documentation on hover

  - fixed by [merlin#1364](https://github.com/ocaml/merlin/pull/1364)
  - fixes duplicate:
    - [ocaml-lsp#344](https://github.com/ocaml/ocaml-lsp/issues/344)
    - [vscode-ocaml-platform#111](https://github.com/ocamllabs/vscode-ocaml-platform/issues/111)

- Correctly rename a variable used as a named/optional argument (ocaml/ocaml-lsp#478)

- When reporting an error at the beginning of the file, use the first line not
  the second (ocaml/ocaml-lsp#489)

### CODE REVIEW

File: packages/ocaml-lsp-server/ocaml-lsp-server.1.8.0/opam [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  "dune" {>= "2.7"}
+  "yojson"
+  "re"
+  "ppx_yojson_conv_lib" {>= "v0.14"}
+  "dune-build-info"
+  "dot-merlin-reader"
+  "pp" {>= "1.1.2"}
+  "csexp" {>= "1.4"}
+  "result" {>= "1.5"}
+  "ocamlformat-rpc-lib" {= "0.18.0"}
```
Comment (kit-ty-kate) [inline packages/ocaml-lsp-server/ocaml-lsp-server.1.8.0/opam:30]:
```suggestion
  "ocamlformat-rpc-lib" {>= "0.18.0" & < "0.19.0"}
```
or
```suggestion
  "ocamlformat-rpc-lib" {>= "0.18.0"}
```
?

**Review by kit-ty-kate — CHANGES_REQUESTED** [review]
```
#=== ERROR while compiling lsp.1.8.0 ==========================================#
# context              2.0.9 | linux/x86_64 | ocaml-base-compiler.4.12.0 | pinned(https://github.com/ocaml/ocaml-lsp/releases/download/1.8.0/jsonrpc-1.8.0.tbz)
# path                 ~/.opam/4.12/.opam-switch/build/lsp.1.8.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p lsp -j 71 @install
# exit-code            1
# env-file             ~/.opam/log/lsp-3305-c61e7b.env
# output-file          ~/.opam/log/lsp-3305-c61e7b.out
### output ###
# File "dune", line 21, characters 5-90:
# 21 |      %{workspace_root}/submodules/dune/otherlibs/stdune-unstable/dune_filesystem_stubs/*.*)
#           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
# Error: Cannot find directory:
# submodules/dune/otherlibs/stdune-unstable/dune_filesystem_stubs
```

File: packages/ocaml-lsp-server/ocaml-lsp-server.1.8.0/opam (hunk 1 above)
Comment (rgrinberg) [reply packages/ocaml-lsp-server/ocaml-lsp-server.1.8.0/opam:30]: @voodoos ping

**Comment by rgrinberg** [comment]
I looked at the errors and it looks like we need a lower bound on re. `>= 1.5.0` should work.

**Comment by Kakadu** [comment]
@rgrinberg Should we expect that all new code actions of ocaml-lsp will be available in most of the editors automatically? (I personally more interested in VsCode)

**Comment by kit-ty-kate** [comment]
Thanks! @rgrinberg could you return the fixes to the opam files upstream?

**Comment by rgrinberg** [comment]
>  Should we expect that all new code actions of ocaml-lsp will be available in most of the editors automatically? (I personally more interested in VsCode)

Yes.

> Thanks! @rgrinberg could you return the fixes to the opam files upstream?

Done. Thanks.

**Pull request merged**